File: src/my_py_pkg/my_py_pkg/turtle_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64  # 导入整数消息类型
from geometry_msgs.msg import Twist  # 导入速度消息类型
from turtlesim.msg import Pose  # 导入位姿消息类型

class TurtleControllerNode(Node):
    """ROS 2 基础 Python 节点类"""
    def __init__(self):
        # 初始化父类，设置节点名称
        super().__init__("turtle_controller")
        # 可在此处添加自定义初始化逻辑（如定时器、发布者、订阅者等）
        self.turtle_vel_publisher_ = self.create_publisher(Twist, "turtle1/cmd_vel", 10)
        self.turtle_pose_subscriber_ = self.create_subscription(
            Pose, "turtle1/pose", self.pose_callback, 10
        )

        self.get_logger().info("TurtleControllerNode has been started.")

    def pose_callback(self, pose: Pose):
        msg = Twist()
        if pose.x < 5.0:
            msg.linear.x = 1.0
            msg.angular.z = 1.0
        else:
            msg.linear.x = 10.0
            msg.angular.z = 5.0
        self.turtle_vel_publisher_.publish(msg)

        self.get_logger().info(f"Published velocity: {msg.linear.x}, {msg.angular.z}")

def main(args=None):
    """节点主函数"""
    # 初始化 rclpy
    rclpy.init(args=args)
    # 创建节点实例
    node = TurtleControllerNode()
    # 保持节点运行（阻塞式）
    rclpy.spin(node)
    # 关闭 rclpy，释放资源
    rclpy.shutdown()

if __name__ == "__main__":
    # 脚本直接运行时执行主函数
    main()

File: src/my_cpp_pkg/src/turtle_controller.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"

class TurtleControllerNode : public rclcpp::Node
{
public:
    /** ROS 2 基础 C++ 节点构造函数 */
    TurtleControllerNode() : Node("turtle_controller")
    {
        tur_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);
        pose_vel_sub_ = this->create_subscription<turtlesim::msg::Pose>(
            "turtle1/pose", 10,
            std::bind(&TurtleControllerNode::pose_vel_callback, this, std::placeholders::_1));
        RCLCPP_INFO(this->get_logger(), "Turtle controller node has been started.");
    }

private:
    /** 发布者 */
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr tur_vel_pub_;
    /** 订阅者 */
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_vel_sub_;
    /** 订阅者回调函数 */
    void pose_vel_callback(const turtlesim::msg::Pose::SharedPtr pose)
    {
        auto vel_msg = geometry_msgs::msg::Twist();
        if (pose->x<5.0){
            vel_msg.linear.x = 1.0;
            vel_msg.angular.z = 1.0;
        }
        else{
            vel_msg.linear.x = 2.0;
            vel_msg.angular.z = 2.0;
        }

        tur_vel_pub_->publish(vel_msg);
    }
};

int main(int argc, char** argv)
{
    // 初始化 rclcpp
    rclcpp::init(argc, argv);
    // 创建节点智能指针实例（ROS 2 推荐使用智能指针）
    auto node = std::make_shared<TurtleControllerNode>();
    // 保持节点运行（阻塞式）
    rclcpp::spin(node);
    // 关闭 rclcpp，释放资源
    rclcpp::shutdown();
    return 0;
}
